R/load.cran.toc.R

```diff
@@ -58,85 +58,35 @@ load.cran.toc <- function(update.toc = FALSE) {
+    dl_toc <- try(download.file(paste0(groundhogR.url, "cran.toc.rds"), toc.path))
 
-      # save to local drive
-      saveRDS(cran.toc, file = toc.path, version = 2, compress = "xz")
-    } # End 3.3.5 - if succeeded at downloading file from website
-
-
-    # UPDATE TIMES
-    # 3.3.6 Try updating times by downloading additional rows from groundhogR server
-    add.times <- try(read.csv(paste0(groundhogR.url, "/differential.times.php?current_date=", max.existing.times.date)))
-
-
-    # 3.3.7 If sucess loading URL
-    if (is.data.frame(add.times)) {
-      # Get pkg_vrs for packages to add
-      add.times.pkg_vrs <- paste0(add.times$Package, "_", add.times$Version)
-
-      # Drop repeated rows in URL and existing (there is the last day of partial to full overlap)
-      add.times.net <- add.times[!add.times.pkg_vrs %in% existing.times.pkg_vrs, ]
-
-      # Add net
-      cran.times <- rbind(existing.times, add.times.net)
-      .pkgenv[["cran.times"]] <- cran.times # Save cran.times to environemnt
+    cran.times <- readRDS(times.path)
+    cran.times$update.date <- as.DateYMD(cran.times$update.date)
 
-      # save to local drive
-      saveRDS(cran.times, file = times.path, version = 2, compress = "xz")
-    } # End 3.3.5 - if succeeded at downloading file from website
+    cran.toc <- readRDS(toc.path)
+    cran.toc$Published <- as.DateYMD(cran.toc$Published)
 
-    # Feedback to user on existing cran.toc
-    message2()
-    message1(
-      "This computer had a database with a list of all versions available for each CRAN package up to ",
-      max.existing.toc.date + 2, " for a total of N=", nrow(existing.toc), " package versions."
-    ) # Add back the two days we took out
-    if (is.data.frame(add.toc)) {
-      message1(
-        "We checked for additions to CRAN since then, and added ",
-        nrow(add.toc.net), " additional entries to the list.\n"
-      )
-    } else {
-      message1("We tried to update till today but it did not work")
-    }
+    .pkgenv[["cran.times"]] <- cran.times
+    .pkgenv[["cran.toc"]] <- cran.toc
 
-    message1("The file with the list is stored here: ", toc.path, "\n-------------------------------")
+  #   # Feedback to user on existing cran.toc
+  #   message2()
+  #   message1(
+  #     "This computer had a database with a list of all versions available for each CRAN package up to ",
+  #     max.existing.toc.date + 2, " for a total of N=", nrow(existing.toc), " package versions."
+  #   ) # Add back the two days we took out
+  #   if (is.data.frame(add.toc)) {
+  #     message1(
+  #       "We checked for additions to CRAN since then, and added ",
+  #       nrow(add.toc.net), " additional entries to the list.\n"
+  #     )
+  #   } else {
+  #     message1("We tried to update till today but it did not work")
+  #   }
+  #
+  #   message1("The file with the list is stored here: ", toc.path, "\n-------------------------------")
```

Comment:
@urisohn, I commented out this part since it doesn't make sense if we're not doing differential updates anymore. Could you please have a look and edit the messages as you see fit?

Reply:
i know you were going to do a time test, what are the results? (comparing incremental update vs downloading the entire .rds from scratch? 

 I am not too worried about speed, again, when updating something a few seconds is not a big deal to users.

I am a bit worried about the implications for traffic volume. If many people use groundhogR (I have no idea what a plausible ceiling on # of users could be), and they each download a 1.5Mb file many times a week...

My intuition is that it will be ok. Not more than 10-20k people will use it, they won't upload more than once a month, it will be fine, and probably well below this. But say it explodes and gets 200,000 user downloading multiple times a week.  At least I would like to think of a calibration and think ahead, if this becomes very successful, is it going to over-burden a dedicated server?  

Somethign like "if this became as big as tidyverse, and people used it everyday, and updated it every other day, it would cost €25.32 a month, relax". 

Or at least to in the back of our minds have a backup strategy, "if this takes off like crazy, here is the mirror server system we could use." Just to put my mind at ease.

Reply:
The results are [here](https://github.com/Bisaloo/groundhogR/issues/17#issuecomment-612468122).

Basically, the differential update is 2x longer in my example (10s vs 5s), averaged on ten runs (not a great sample size but I didn't want to put to much stress on the server).

> I am a bit worried about the implications for traffic volume. If many people use groundhogR (I have no idea what a plausible ceiling on # of users could be), and they each download a 1.5Mb file many times a week...

This is unlikely, even with many users since they will have a version of `cran.toc.rds` and `cran.times.rds` when they download groundhogR from CRAN. They will only fetch it from the server when an update if necessary (`update_cran.toc_if.needed()` is `TRUE`), which should not happen that often.

Additionally, beyond what happens on the user side, there is a bandwith / processing power issue for the server. You php script will take some amount of processing power to run and if the number of users blows up, I suspect bandwith is cheaper than CPU (but please let me know if I'm wrong here).

Reply:
OK, i am not sure what's optimal, but, it seems easy to scale up storage and bandwidth and harder to scale up processing, so let's go with downloading the entire file each time.